pandas/core/algorithms.py

```diff
@@ -114,11 +114,8 @@ def _ensure_data(
         values = extract_array(values, extract_numpy=True)
 
     # we check some simple dtypes first
-    if is_object_dtype(dtype):
+    if is_object_dtype(dtype) or (is_object_dtype(values) and (dtype is None)):
```

Comment:
can you remove the parens around dtype is None as confusing (and actually put around the entire expression and format)